### ISSUE TITLE
tools/master-ip-setup stalls on Ubuntu 13.10

### ISSUE THREAD

**Issue by Ganeti-Issues-Migrator** [ganeti/ganeti#686]
Originally reported of Google Code with ID 630.

```
On Ubuntu 13.10 (Saucy Salamander), ganeti (both 2.7.0 and 2.9.1) stalls on /usr/lib/ganeti/tools/master-ip-setup start/stop. 

The reason is that fping does not exit when the target is unreachable. In other words, if the master ip is not assigned fping stalls for ever.

# fping -S 127.0.0.1 192.168.0.225

192.168.0.225 error while sending ping: Invalid argument

192.168.0.225 error while sending ping: Invalid argument

192.168.0.225 error while sending ping: Invalid argument

192.168.0.225 error while sending ping: Invalid argument

192.168.0.225 error while sending ping: Invalid argument

...


I have a simple fix for this that works on Ubuntu Saucy and Debian Wheezy. We can add "-c 1" to all fpings to make sure it does not stall. 

This the patch:

diff --git a/tools/master-ip-setup b/tools/master-ip-setup
index 9716bc3..717abad 100755
--- a/tools/master-ip-setup
+++ b/tools/master-ip-setup
@@ -39,13 +39,13 @@ start() {
   esac
 
   # Check if the master IP address is already configured on this machine
-  if fping -S 127.0.0.1 $MASTER_IP >/dev/null 2>&1; then
+  if fping -c 1 -S 127.0.0.1 $MASTER_IP >/dev/null 2>&1; then
     echo "Master IP address already configured on this machine. Doing nothing."
     exit 0
   fi
 
   # Check if the master IP address is already configured on another machine
-  if fping $MASTER_IP >/dev/null 2>&1; then
+  if fping -c 1 $MASTER_IP >/dev/null 2>&1; then
     echo "Error: master IP address configured on another machine." >&2
     exit 1
   fi
@@ -66,7 +66,7 @@ stop() {
   if ! ip addr show dev $MASTER_NETDEV | \
      grep -F " $MASTER_IP/$MASTER_NETMASK" >/dev/null 2>&1; then
     # Check if the master IP address is configured on a wrong device
-    if fping -S 127.0.0.1 $MASTER_IP >/dev/null 2>&1; then
+    if fping -c 1 -S 127.0.0.1 $MASTER_IP >/dev/null 2>&1; then
       echo "Error: master IP address configured on wrong device," \
            "can't shut it down." >&2
       exit 1


```
Originally added on 2013-11-30 01:14:09 +0000 UTC.

Attachments: 
[master-ip-setup.patch](https://storage.googleapis.com/google-code-attachments/ganeti/issue-630/comment-0/master-ip-setup.patch)

**Comment by Ganeti-Issues-Migrator** [comment]
```
And btw it'd be awesome, if it could be cherry picked for 2.7.0.

People have reported this problem on ubuntu 13.10:
https://groups.google.com/forum/#!topic/ganeti/0vbFRuhk_II
```
Originally added on 2013-11-30 01:17:40 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Hi,

Thank you for your report and patch. While we review the case, please take the time to sign the Contribute License Agreement if you haven't done so already:

https://developers.google.com/open-source/cla/individual

We cannot otherwise accept your patch.
```
Originally added on 2013-11-30 13:47:46 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Sure, I just signed and submitted the CLA. 

```
Originally added on 2013-11-30 16:13:32 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Confirmed. This seems to be a bug in fping 3.5, because it's manual page says that

  -c n .... count of pings to send to each target (default 1)

Thanks for reporting it.
```
Originally added on 2013-12-02 08:23:12 +0000 UTC.

Changed State: Started

**Comment by Ganeti-Issues-Migrator** [comment]
```
-- Empty comment --
```
Originally added on 2013-12-02 08:23:26 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Fixed in:

commit 71b75e7a7e3876a21c2f54bff1a45d4de65f79e0
Author: Petr Pudlak <pudlak@google.com>
Date:   Mon Dec 2 11:04:54 2013 +0100

    Add a packet number limit to "fping" in master-ip-setup
    
    This fixes issue #630. Apparently there is a bug in fping 3.5 where it
    loops forever without "-c" given an unreachable host, even though
    "-c 1" should be the default according to the man page.
    
    The "-c" flag works on Squeeze. Checking the man pages on the Internet,
    fping supported "-c" at least since 2007. So there should be no backward
    compatibility problems.
    
    Signed-off-by: Petr Pudlak <pudlak@google.com>
    Reviewed-by: Jose A. Lopes <jabolopes@google.com>

```
Originally added on 2013-12-02 13:56:53 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
The bug has been fixed in 2.8, which will get a new release soon, and eventually will get merged into 2.9 and up.

The fix has also been cherry-picked into 2.7, but it doesn't justify a new 2.7 release. However, if there will be a new maintenance release of 2.7 in the future, it'll be included there.
```
Originally added on 2013-12-03 11:31:05 +0000 UTC.

Changed State: Fixed
Added to Milestone: Release2.8

**Comment by Ganeti-Issues-Migrator** [comment]
```
-- Empty comment --
```
Originally added on 2013-12-13 16:18:23 +0000 UTC.

Changed State: Released
